src/smtp.rs

```diff
@@ -220,3 +224,231 @@ impl Smtp {
+        Err(crate::smtp::send::Error::Envelope(err)) => {
+            // Local error, job is invalid, do not retry.
+            smtp.disconnect().await;
+            warn!(context, "SMTP job is invalid: {}", err);
+            Status::Finished(Err(err.into()))
+        }
+        Err(crate::smtp::send::Error::NoTransport) => {
+            // Should never happen.
+            // It does not even make sense to disconnect here.
+            error!(context, "SMTP job failed because SMTP has no transport");
+            Status::Finished(Err(format_err!("SMTP has not transport")))
+        }
+        Err(crate::smtp::send::Error::Other(err)) => {
+            // Local error, job is invalid, do not retry.
+            smtp.disconnect().await;
+            warn!(context, "unable to load job: {}", err);
+            Status::Finished(Err(err))
+        }
+        Ok(()) => Status::Finished(Ok(())),
+    };
+
+    if let Status::Finished(Err(err)) = &status {
+        // We couldn't send the message, so mark it as failed
+        message::set_msg_failed(context, msg_id, Some(err.to_string())).await;
+    }
+    status
+}
+
+/// Sends message identified by `smtp` table rowid over SMTP connection.
+///
+/// Removes row if the message should not be retried.
+pub(crate) async fn send_msg_to_smtp(
+    context: &Context,
+    smtp: &mut Smtp,
+    rowid: i64,
+) -> anyhow::Result<()> {
+    if let Err(err) = smtp
+        .connect_configured(context)
+        .await
+        .context("SMTP connection failure")
+    {
+        smtp.last_send_error = Some(format!("SMTP connection failure: {:#}", err));
+        return Err(err);
+    }
+
+    let (body, recipients, msg_id) = context
+        .sql
+        .query_row(
+            "SELECT mime, recipients, msg_id FROM smtp WHERE id=?",
+            paramsv![rowid],
+            |row| {
+                let mime: Vec<u8> = row.get(0)?;
+                let recipients: String = row.get(1)?;
+                let msg_id: MsgId = row.get(2)?;
+                Ok((mime, recipients, msg_id))
+            },
+        )
+        .await?;
+    let recipients_list = recipients
+        .split(' ')
+        .filter_map(
+            |addr| match async_smtp::EmailAddress::new(addr.to_string()) {
+                Ok(addr) => Some(addr),
+                Err(err) => {
+                    warn!(context, "invalid recipient: {} {:?}", addr, err);
+                    None
+                }
+            },
+        )
+        .collect::<Vec<_>>();
+
+    let status = smtp_send(context, recipients_list, body, smtp, msg_id, rowid).await;
+    match status {
+        Status::Finished(res) => {
+            if res.is_ok() {
+                msg_id.set_delivered(context).await?;
+
+                context
+                    .sql
+                    .execute("DELETE FROM smtp WHERE id=?", paramsv![rowid])
+                    .await?;
+
+                context.emit_event(EventType::DeletedBlobFile(format!("SMTP row {}", rowid)));
```

Comment:
This is for compatibility with existing tests using this event as an indicator that message has been successfully sent. It can be interpreted as "blob has been removed from the database".

Reply:
if this is purely for tests i wonder if we should fix up the tests instead.  I'd be happy to leave a comment here to that effect and leave it for now, but it seems weird to emit such an even only to make tests happy.

Reply:
what do you think?

Reply:
I removed this line and changed the python test relying on it.